### DISCUSSION TITLE
How can I display a shape on demand using a predefined driver?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#6870] [Q&A] [answered]
I am trying to display a shape only if the model is allowing to show it. For example I have `FancyFormPart` with a `bool` property called `ShowTypes`. The `FancyFormPart` has its own driver. If the property is set to true, I want to render a part called `PropertyTypePart`. Note that `PropertyTypePart` has its own driver also.

I am puzzled on how can I display a shape on demand like that? My view would look something like this 

```
@model FancyFormPartViewModel

@if(Model.ShowTypes)
{
    // How can I create the shape of the PropertyTypePart to allow the `DisplayAsync()` method to display it here??
    @await DisplayAsync(....)
}
```

**Comment by Piedone** [comment, marked as answer]
A basic idea in Orchard is that content types are a composition of independent parts. It seems to me that you're attempting to make one part depend on another one. Since you can't really rely on parts being evaluated (like displayed) in a particular order, I encourage you to change how you approach this task. Can you tell what exactly are you trying to achieve?

That being said, here are a few ideas:

- `Display()` in the driver can return `null` based on the part's content or anything else.
- Instead of trying to render `PropertyTypePart`'s shape in `FancyFormPart` you can render an ad-hoc shape there directly (if you want to factor that piece of markup out). 
- You can just override the whole `Content` shape of that content item and render everything there, instead of relying on each part to do their job (this is suitable only if this is for a very particular use-case).
- You can even prevent shapes from being rendered in an `IShapeDisplayEvents` implementation.

Also, you don't have to (or rather, shouldn't try to) explicitly build the display of `PropertyTypePart` in `FancyFormPart`'s shape template. It's the Display Manager's job to compose the content item's output from all the content part drivers' shapes' renderings (and else).

**Comment by MikeAlhayek** [comment]
I have a content type called `FancyPage`. this type will show a form "each input will have a boolean field that will indicate if it should be visible or not.

This form will be handled using a single page approach. So when the user submit the form, we'll submit the form using AJAX request and populate the page accordingly. There will be many content-items of the fancypage. The main difference will be what inputs to show on that form.

Since I am unable to render one part from another. I should be able to do the same thing from the content-type, right?

Then am I able to compose a shape on the fly from my `Content-FancyPage.cshtml`? I mean I can add boolean fields like `ShowTypes` to the `FancyPage` type. But again, how would I compose the shape on demand?

**Reply by Piedone** [reply]
I see. In that case, is there anything that would prevent you from keeping the show indicator and the rest in a single part?

**Reply by MikeAlhayek** [reply]
No. I can either create a part with multiple boolean fields "the flags (i.e., `ShowType`)" or I can just make the flags part of the `FancyPage` type.

**Reply by Piedone** [reply]
I'd do that if I were you then. Or, you can create separate parts for a logical subset of the fields that somehow belong together, or create a content field for every kind of field, or create a widget for every field that you'll then throw together in a Flow Part. Just some ideas.

**Reply by MikeAlhayek** [reply]
I am okay with create a part that has all of the flags (i.e., `FancyFormFlagsPart`). But, how can I compose a shape on demand.

Here is the view of the content's type  `Content-FancyPage.cshtml`

```
@model FancyPageViewModel
@{
    var flags = (FancyFormFlagsPart)Model.ContentItem.Content.FancyFormFlagsPart;
}

@if (flags.ShowType)
{
    // At this point, we know that the that we need to compose the `PropertyTypePart` on demand.
    // How can I create the shape of the `PropertyTypePart` to allow the `DisplayAsync()` method to display it here??
    @await DisplayAsync(....)
}
```

**Reply by Piedone** [reply]
Just some clarification: You're indeed composing a shape on demand but not `PropertyTypePart`. So, you want to display an ad-hoc shape. The Training Demo Module has some thoroughly explained samples on this, you can start from here: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Map.cs#L142

**Reply by MikeAlhayek** [reply]
> You're indeed composing a shape on demand but not PropertyTypePart

I am not sure why wouldn't the part be `PropertyTypePart`? 

Just to make sure I understand what I should be doing. In the view I would do something like `@await DisplayAsync(Model.Type)` to render the "Type" shape. But to compose the `Type` I would so that using a filter, am I understanding the instructions correct? If that is that case, how can I check if `filterContext.Result` is a content type of `FancyPage` also, how can access the `FancyFormFlagsPart` to know if I need to compose the specific shape that needs to be injected?

What shape does [this line](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/b9cdc1728949515c7dd28db387513d60a4d0c66e/Filters/ShapeInjectionFilter.cs#L55) inject into the Content zone?

**Reply by Piedone** [reply]
If `PropertyTypePart` is indeed a part, then it's something that you'd use to compose the content item, not something that you'd display ad-hoc.

I just noticed that I didn't link to the sample I thought about, check out this one instead, this is for exactly your scenario and will make a lot of things clearer: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/b9cdc1728949515c7dd28db387513d60a4d0c66e/Controllers/DisplayManagementController.cs#L33 No filter needed in your case, that's another way to do things but not relevant here.

I also clarified the sample you've linked to: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Filters/ShapeInjectionFilter.cs#L54

**Comment by jtkech** [comment]
As @Piedone said there are many way to do it, e.g. in an overridden template of your parent content type, you can render what you want, we have helpers to retrieve a given part and then check its properties e.g. to not render another part.

Hmm, but as @Piedone said maybe better to do it through display drivers, e.g. part1 driver would provide on display its related shape and based on a condition also combine the shape of part2, part2 driver would provide nothing on display.

But better to let part2 driver decide itself to provide or not its related shape on display

        public override IDisplayResult Display(MySecondPart mySecondPart, BuildPartDisplayContext context)
        {
            if (mySecondPart.ContentItem.As<MyFirstPart>()?.SomeBoolProperty == false)
            {
                return null;
            }

**Reply by Piedone** [reply]
Ah and one more thing @CrestApps that occurred to me from Jean-Thierry's answer: You can also return multiple shapes from a single `Display()` if you want to organize things a bit. See: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Drivers/BookDisplayDriver.cs#L19 (this is not a content part driver but a simple entity one but this works the same).

**Comment by ns8482e** [comment]
@CrestApps design your Parts as reusable piece of block. 

If your one part is depends on another part - then you can’t use one part without another - not a good design.

Keep `ShowTypes ` and `PropertyTypePart` in same part under same model - in display driver return shape if `ShowTypes` is true.

**Comment by weirdyang** [comment]
or you can do something like this:
```

@(if showType){
 
    @await DisplayAsync(await New.MyShape(Foo: 1, Bar: "a"))
}
```
this could work too
